[feature] Support traffic direction filter autocompletion

silvanbitterli opened the pull request:
This PR provides support for autocompletion of the `{dir|direction}` keyword. 
As the `{dir|direction}` keyword imposes some conditions on the structure of the condition string, we want to ensure that:
- keywords are only suggested when they are applicable
- condition strings are automatically terminated if any additional condition would yield an invalid query

Examples: 
- `dir` is suggested for an empty condition string as well as for `(sip = 127.0.0.1 & port = 22) & ` but not for `sip = 127.0.0.1 & port = 22 & ` (`dir`wouldn't occur at the top-level in this case)
- `sip = 127.0.0.1 & dir = inb` yields the automatically terminated condition string `sip = 127.0.0.1 & dir = inbound`
- `dir = in & sip = 127.0.0.1` is automatically terminated

As this PR contains complex filtering logic, we have added extensive comments, describing each filter.

basman commented on cmd/goquery_completion/conditional.go:
```diff
@@ -169,3 +224,151 @@ func conditional(args []string) []string {
+	// directly after a top level condition with non-empty value,
+	// the condition string must end.
+	// The condition value is only validated for direction filters.
+	// Note: this case handles termination in case the
```
Insert an empty comment line above Note:

basman commented on cmd/goquery_completion/conditional.go:
```diff
@@ -169,3 +224,151 @@ func conditional(args []string) []string {
+	// If there is exactly one suggestion, and the current
+	// condition is a direction filter and this suggestion is
+	// exactly what the user specified, the condition string must end.
+	// Note: this handles termination of condition strings ending
```
Insert an empty comment line above Note:

basman commented on cmd/goquery_completion/conditional.go:
```diff
@@ -169,3 +224,151 @@ func conditional(args []string) []string {
+	// (e.g. terminates "dir = in & sip = 127.0.0.1", BUT also
+	// "dir = in & sip = 42")
+	if len(suggs) == 0 && dirKeywordOccurrences > 0 && openParens == 0 && nTokens > 3 {
+
```
remove empty leading line (wherever you open a block { )

els0r replied:
I'm with @basman here. Mainly because `golanci-lint` drilled this into my head at an early stage when I switched to VSCode.

Now that we have ourselves a little community, let us "formalize" some guidelines. Not that it must happen this PR, but in general, I propose:

* `golanci-lint` default settings as the linter base
* no unnamed returns. Even if return parameters are named in function definition
* no empty line between `{` and the next statements
* every log line starts the message _lowercase_
* comments: no preference as long as they don't collide with vet/lint. What I gravitate towards is more often than not a _lowercase_ letter at the start of an in-scope comment

Happy to put these down in writing once we converge on the above

silvanbitterli replied:
The guidelines seem very reasonable to me. 
I personally prefer an extra line in case the block starts with a comment as @fako1024 proposed, but if the default `golangci-lint` suggests to remove it, it's also fine with me.
With regards to comments my preference is to differentiate between single-sentence and multi-sentence comments. For single-sentence comments I prefer lowercase start without trailing dot, while for multi-sentence comments I prefer to stick with regular grammar rules. But I don't have a strong opinion on this.

fako1024 replied:
I'd lean towards @basman's opinion about comments following regular grammar (and, as stated before, leaving an empty line to the comment at the beginning of a block, which doesn't / shouldn't conflict with `golangci-lint`, I've been doing it all along), i.e. starting with an uppercase letter - always, even for a short comment (mostly for the reason of consistency / simplicity of rule, IMHO it unnecessarily complicates things if you have two different rules for virtually the same thing) - after all, a sentence is a sentence is a sentence. I'm not too pedantic about the dot though in the single sentence case...

As for the rest of the proposal I'm totally game (can't promise I won't forget a named return here and there though :see_no_evil:), thanks @els0r ...

basman commented on cmd/goquery_completion/conditional_test.go:
```diff
@@ -0,0 +1,85 @@
+		suggs := conditional(test.in)
+		nSuggs := len(suggs)
+		require.Equalf(t, test.expectedNSuggestions, nSuggs, "Expected: %d Got: %d", test.expectedNSuggestions, nSuggs)
+
```
remove empty line

silvanbitterli commented:
If you haven't yet started reviewing, please wait until my next commit.

fako1024 commented on cmd/goquery_completion/conditional.go:
```diff
@@ -31,6 +31,86 @@ func openParens(tokens []string) int {
 }
 
+// dirKeywordCount returns the number of dir keyword occurrences in tokens
+func dirKeywordCount(tokens []string) int {
```
As hinted in the comment in the draft (sorry, didn't see it was still in that state, I just did what was asked in the notification and blindly reviewed :stuck_out_tongue: ), these counting functions are prime candidates for named return parameters, because it unclutters the code (and removes the need to manually initialize a local variable). Also, IMHO if the (short) function is already clearly named and just has a single counter, the variable name can easily be shortened (e.g. to `count` in this case):
```
func dirKeywordCount(tokens []string) (count int) {
	for _, token := range tokens {
		if token == types.FilterKeywordDirection || token == types.FilterKeywordDirectionSugared {
			count++
		}
	}
	return
}
```

fako1024 commented on cmd/goquery_completion/conditional.go:
```diff
@@ -31,6 +31,86 @@ func openParens(tokens []string) int {
+// unnestedBinaryOpOccurrences returns the position of the first 'unnested' binary logical
+// operator inside tokens and the respective token.
+// It returns -1 and an empty string if tokens do not contain any binary logical operator.
+func unnestedBinaryOpOccurrences(tokens []string) int {
```
See comment above about named return parameters.

fako1024 commented on cmd/goquery_completion/conditional.go:
```diff
@@ -31,6 +31,86 @@ func openParens(tokens []string) int {
+// firstUnnestedBinaryOp returns the position of the first 'unnested' binary logical
+// operator inside tokens and the respective token.
+// It returns -1 and an empty string if tokens do not contain any binary logical operator.
+func firstUnnestedBinaryOp(tokens []string) (int, string) {
```
Here I'd keep it because your fallback return value isn't the null element of at least one of the parameters...

fako1024 commented on cmd/goquery_completion/conditional.go:
```diff
@@ -169,3 +285,127 @@ func conditional(args []string) []string {
+// has already appeared, ensuring that the dir keyword only occurs at valid places.
+func verifySuggestion(sugg suggestion, tokens []string, prev string, openParens, dirKeywordCount int) bool {
+	if strings.Contains(sugg.token, types.FilterKeywordDirection) ||
+		strings.Contains(sugg.token, types.FilterKeywordDirectionSugared) ||
```
Nitpick: Is this alignment correct? AFAIK this usually gets aligned with the first condition, not with the inner block... :thinking:

silvanbitterli replied:
I also always thought this alignment looked weird but I think it is correct (`gofmt` does not change it).

fako1024 replied:
Alright then, who wants to argue with `gofmt`... :stuck_out_tongue: - Thanks for cross-checking!

fako1024 commented on cmd/goquery_completion/conditional.go:
```diff
@@ -169,3 +285,127 @@ func conditional(args []string) []string {
+		for _, direction := range types.DirectionFilters {
+			// if dir is specified on the right side of a top-level conjunction,
+			// do not allow any trailing tokens
+			if prev == string(direction) && last == "" || last == string(direction) {
```
Just an idea: You could change the definition of the `FilterTypeDirection` in `types.go` to:
```
// FilterTypeDirection denotes filters wrt. directionality of traffic
type FilterTypeDirection = string
```

That way, you could skip the casts for comparisons everywhere. On the other hand you'd weaken / loosen the typing, but having to cast for string comparisons is often an indicator that you actually want the type equality... But as I said, just an idea...

silvanbitterli replied:
I think for now your suggestion is the cleanest solution and I adjusted the code accordingly. 
In case other types of filters are needed in the future, it might be nice to separate the different kinds of filters via the type system.

fako1024 reviewed (approved):
A few nitpicks / suggestions from my side, but apart from that: LGTM! :tada:

basman commented on cmd/goquery_completion/conditional.go:
```diff
@@ -31,6 +31,29 @@ func openParens(tokens []string) int {
 }
 
+// dirKeywordOccurrences returns the number of dir keyword occurrences in tokens
+func dirKeywordOccurrences(tokens []string) int {
```
Implicit (=shortened) return statements make the code harder to read. But it is your code. Let's see, what els0r has to say.

els0r replied:
See previous comment about guidelines. Even though it sometimes seems pedantic, I'm leaning towards named returns. Even if the return parameter is already named.

els0r reviewed (approved):
LGTM. Looking forward to using this!

basman commented on cmd/goquery_completion/conditional.go:
```diff
@@ -152,11 +251,28 @@ func conditional(args []string) []string {
-				suggs = append(suggs, sugg)
+		prevprev := antepenultimate(tokens)
+		prev := penultimate(tokens)
+		openParens := openParens(tokens)
```
Move this line down to 261, so we avoid searching the entire token list in case of an invalid condition string.

silvanbitterli commented:
Thanks for your feedback and support! 
I have now addressed all the remaining issues and also double checked that this PR will fix the previously introduced `goimports` regression.
I will proceed to squash and merge.
